File: sdk/ctv/android-tv/src/main/kotlin/com/rivalapexmediation/ctv/ads/InterstitialAd.kt
package com.rivalapexmediation.ctv.ads

import android.app.Activity
import androidx.media3.ui.PlayerView
import com.rivalapexmediation.ctv.ApexMediation
import com.rivalapexmediation.ctv.network.AuctionWin
import com.rivalapexmediation.ctv.render.Beacon
import com.rivalapexmediation.ctv.render.VideoRenderer
import com.rivalapexmediation.ctv.util.Logger

class InterstitialAd(private val placementId: String) {
    @Volatile private var win: AuctionWin? = null
    @Volatile private var loaded = false

    fun load(floorCpm: Double = 0.0, callback: AdLoadCallback) {
        if (!ApexMediation.isInitialized()) { callback.onError("not_initialized"); return }
        val consent = ApexMediation.consent()
        ApexMediation.client().requestBid(placementId, "interstitial", floorCpm, consent) { result ->
            if (result.noFill) { ApexMediation.postMain { callback.onError("no_fill") }; return@requestBid }
            if (result.error != null) { ApexMediation.postMain { callback.onError(result.error) }; return@requestBid }
            win = result.win
            loaded = win != null
            if (loaded) ApexMediation.postMain { callback.onLoaded() } else ApexMediation.postMain { callback.onError("no_fill") }
        }
    }

    fun isReady(): Boolean = loaded && win != null

    /**
     * Show the interstitial into a provided PlayerView (full-screen in host Activity recommended).
     */
    fun show(activity: Activity, playerView: PlayerView, callback: AdShowCallback) {
        val w = win
        if (!isReady() || w == null) { callback.onError("not_ready"); return }
        val renderer = VideoRenderer(activity)
        renderer.attach(playerView)
        try {
            // Fire impression when first frame is rendered
            renderer.play(w.creativeUrl, onReady = {
                try { Beacon.fire(w.tracking.impression) } catch (_: Throwable) {}
                ApexMediation.postMain { callback.onShown() }
            }, onEnded = {
                ApexMediation.postMain { callback.onClosed() }
                renderer.release()
            })
        } catch (e: Exception) {
            Logger.w("Interstitial show error: ${e.message}", e)
            callback.onError("render_error")
        }
    }

    /**
     * Report a user click for this ad (fires signed click beacon).
     */
    fun reportClick() {
        val w = win ?: return
        try { Beacon.fire(w.tracking.click) } catch (_: Throwable) {}
    }
}

File: sdk/ctv/tvos/Sources/CTVSDK/InterstitialAd.swift
import Foundation
import AVKit

public final class InterstitialAd {
    private let placementId: String
    private var win: AuctionWin?
    private var loaded = false

    public init(placementId: String) { self.placementId = placementId }

    public func load(floorCpm: Double = 0.0, completion: @escaping (_ error: String?) -> Void) {
        let sdk = ApexMediation.shared
        guard sdk.isInitialized else { completion("not_initialized"); return }
        if let reason = sdk.loadGuard(for: placementId) { completion(reason); return }
        sdk.client.requestBid(placementId: placementId, adFormat: "interstitial", floorCpm: floorCpm, consent: sdk.consent) { res in
            if let error = res.error {
                completion(error.reason)
                return
            }
            guard let win = res.win else {
                completion("no_fill")
                return
            }
            AdCache.shared.store(win: win, for: self.placementId)
            self.win = win
            self.loaded = true
            completion(nil)
        }
    }

    public var isReady: Bool { AdCache.shared.peek(placementId: placementId) != nil }

    public func show(on viewController: UIViewController, completion: @escaping (_ error: String?) -> Void, closed: @escaping () -> Void) {
        guard let cachedWin = AdCache.shared.take(placementId: placementId) else {
            if loaded {
                loaded = false
                completion("expired")
            } else {
                completion("not_ready")
            }
            return
        }
        win = cachedWin
        loaded = false
        if let reason = ApexMediation.shared.showGuard(for: placementId) { completion(reason); return }
        let player = AVPlayer(url: URL(string: cachedWin.creativeUrl)!)
        let controller = AVPlayerViewController()
        controller.player = player
        controller.modalPresentationStyle = .fullScreen
        let tracker = VideoTracker(tracking: cachedWin.tracking)
        viewController.present(controller, animated: true) {
            tracker.attach(to: player)
            Beacon.fire(cachedWin.tracking.impression, eventName: "impression")
            controller.player?.play()
            completion(nil)
        }
        var observer: NSObjectProtocol?
        observer = NotificationCenter.default.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: player.currentItem, queue: .main) { _ in
            if let obs = observer { NotificationCenter.default.removeObserver(obs) }
            tracker.markComplete()
            tracker.markClose()
            controller.dismiss(animated: true) {
                tracker.stop()
                closed()
            }
        }
    }

    /// Report a user click for this ad (fires signed click beacon).
    public func reportClick() {
        guard let w = win else { return }
        Beacon.fire(w.tracking.click, eventName: "click")
    }
}
